### PR TITLE
perf(tree): avoid cloning change enricher checkout on every change

### PR DESCRIPTION
**Pull request by yann-achard-MS**
## Description

Before this PR, commits were enriched between their application to the local checkout and their sending to the service.
This required keeping a defensive clone of the enricher checkout so that it could enrich a commit based on the state immediately before that commit was applied.

This PR change the old order of operations (apply-enrich-submit) to enrich-apply-submit. This allows the enrichment to be performed based on the forest and detached field index of the top-level checkout instead of a clone, thus making cloning necessary only in resubmit scenarios.

While the above sounds simple, it prompted a lot of changes to accommodate the intricacies of transactions.
A subsequent PR (perhaps after GA if it is deemed to destabilizing) could clean up a lot of this complexity by making transactions less special throughout the system.

## Breaking Changes

None

### CODE REVIEW

File: packages/dds/tree/src/shared-tree-core/sharedTreeCore.ts [hunk 1]
```diff
@@ -238,7 +276,7 @@ export class SharedTreeCore<TEditor extends ChangeFamilyEditor, TChange> extends
 	 * @returns the submitted commit. This is undefined if the underlying `SharedObject` is not attached,
 	 * and may differ from `commit` due to enrichments like detached tree refreshers.
 	 */
-	protected submitCommit(
+	private submitCommit(
 		commit: GraphCommit<TChange>,
 		schemaAndPolicy: ClonableSchemaAndPolicy,
 		isResubmit = false,
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree-core/sharedTreeCore.ts:282]: Remove return type

File: packages/dds/tree/src/shared-tree-core/branchCommitEnricher.ts [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	/**
+	 * @param commit - A commit previously passed to {@link BranchCommitEnricher.prepareCommit}.
+	 * @returns The enriched commit corresponds to the given commit.
+	 *
+	 * Commits are expected to be retrieved in the same order they were prepared (FIFO).
+	 */
+	public getPreparedCommit(commit: GraphCommit<TChange>): GraphCommit<TChange> {
+		const prepared = this.preparedCommits.shift();
```
Comment (yann-achard-MS) [inline packages/dds/tree/src/shared-tree-core/branchCommitEnricher.ts:74]: Make this a map

File: packages/dds/tree/src/shared-tree-core/branchCommitEnricher.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+/**
+ * Utility for enriching commits from a {@link Branch} before these commits are applied and submitted.
+ */
+export class BranchCommitEnricher<TChange> {
+	private transactionEnricher?: TransactionEnricher<TChange>;
+	private readonly rebaser: ChangeRebaser<TChange>;
+	private readonly enricher: ChangeEnricherReadonlyCheckout<TChange>;
+	/**
+	 * Maps each local commit to the corresponding enriched commit.
+	 * Entries are added then the commits are prepared (before being applied and submitted).
```
Comment (noencke) [inline packages/dds/tree/src/shared-tree-core/branchCommitEnricher.ts:20]:
```suggestion
	 * Entries are added when the commits are prepared (before being applied and submitted).
```

File: packages/dds/tree/src/shared-tree-core/changeEnricher.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+import { RevisionTag } from "../core/index.js";
+import { disposeSymbol } from "../util/index.js";
+
+/**
+ * A checkout that can be used by {@link SharedTreeCore} or {@link DefaultResubmitMachine} to enrich changes with refreshers.
+ *
+ * This is similar to a {@link TreeCheckout} in that it represents the state of the tree at a specific revision.
+ * But unlike a `TreeCheckout`...
+ * - It is not backed by a branch because the `CommitEnricher` that owns it controls which revision it should represent.
+ * - The host application has no knowledge of it, so applying changes to it has not impact on the application.
```
Comment (noencke) [inline packages/dds/tree/src/shared-tree-core/changeEnricher.ts:15]:
```suggestion
 * - The host application has no knowledge of it, so applying changes to it has no impact on the application.
```

**Review by noencke — APPROVED** [review]

**Pull request merged**